crates/apollo-compiler/src/diagnostic.rs

```diff
@@ -0,0 +1,344 @@
+            Ok(source_file.ariadne())
+        } else if *file_id == FileId::NONE || *file_id == FileId::HACK_TMP {
+            static EMPTY: OnceLock<ariadne::Source> = OnceLock::new();
+            Ok(EMPTY.get_or_init(|| ariadne::Source::from("")))
+        } else {
+            Err(Box::new(NotFound(*file_id)))
+        }
+    }
+
+    fn display<'a>(&self, file_id: &'a FileId) -> Option<Box<dyn fmt::Display + 'a>> {
+        if *file_id != FileId::NONE {
+            struct Path(Arc<SourceFile>);
+            impl fmt::Display for Path {
+                fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                    self.0.path().display().fmt(f)
+                }
+            }
+            let source_file = self.0.get(file_id)?;
+            Some(Box::new(Path(source_file.clone())))
+        } else {
+            struct NoSourceFile;
+            impl fmt::Display for NoSourceFile {
+                fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                    f.write_str("(no source file)")
+                }
+            }
+            Some(Box::new(NoSourceFile))
+        }
+    }
+}
+
+/// A pretty-printable diagnostic.
+pub struct Diagnostic<T> {
+    pub sources: SourceMap,
+    pub location: Option<NodeLocation>,
+    pub error: T,
+}
+
+impl<T> std::ops::Deref for Diagnostic<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        &self.error
+    }
+}
+
+impl<T: std::error::Error + ToDiagnostic> std::error::Error for Diagnostic<T> {}
+
+impl<T: ToDiagnostic> ToDiagnostic for &T {
+    fn location(&self) -> Option<NodeLocation> {
+        ToDiagnostic::location(*self)
+    }
+
+    fn report(&self, report: &mut CliReport) {
+        ToDiagnostic::report(*self, report)
+    }
+}
+
+impl<T: ToDiagnostic> Diagnostic<T> {
+    /// Get the line and column number where this diagnostic was raised.
+    pub fn get_line_column(&self) -> Option<GraphQLLocation> {
+        GraphQLLocation::from_node(&self.sources, self.error.location())
+    }
+
+    /// Get a [`serde`]-serializable version of the current diagnostic. The shape is compatible
+    /// with the JSON error shape described in [the GraphQL spec].
+    ///
+    /// [the GraphQL spec]: https://spec.graphql.org/draft/#sec-Errors
+    pub fn to_json(&self) -> GraphQLError
+    where
+        T: ToString,
+    {
+        GraphQLError::new(self.error.to_string(), self.error.location(), &self.sources)
+    }
+
+    /// Produce the diagnostic report, optionally with colors for the CLI.
+    fn report(&self, color: bool) -> CliReport {
+        let mut report =
+            CliReport::builder(self.sources.clone(), self.error.location()).with_color(color);
+        self.error.report(&mut report);
+        report
+    }
+
+    /// Pretty-print the diagnostic to a [`Write`].
+    ///
+    /// [`Write`]: std::io::Write
+    pub fn write(&self, color: bool, w: impl std::io::Write) -> std::io::Result<()> {
```

Comment:
At call sites it’s not obvious what this boolean means. What do you think of `enum Color { None, StderrIsTerminal }` ?